web3/middleware/stalecheck.py

```diff
@@ -0,0 +1,44 @@
+import time
+
+from web3.exceptions import StaleBlockchain
+
+SKIP_STALECHECK_FOR_METHODS = set([
+    'eth_getBlockByNumber',
+])
+
+
+def _isfresh(block, allowable_delay):
+    return block and time.time() - block['timestamp'] <= allowable_delay
+
+
+def make_stalecheck_middleware(allowable_delay=0):
+    '''
+    Use to require that a function will run only of the blockchain is recently updated.
+
+    This middleware takes an argument, so unlike other middleware, you must make the middleware
+    with a method call.
+    For example: `make_stalecheck_middleware(allowable_delay=60*5)`
+
+    If the latest block in the chain is older than 5 minutes in this example, then the
+    middleware will raise a StaleBlockchain exception.
+    '''
+    if not allowable_delay:
+        raise ValueError("You must set a non-zero allowable delay for this middleware")
+
+    def stalecheck_middleware(make_request, web3):
+        cache = {'latest': None}
```

Comment:
Since we only ever access the key `'latest'` why not just store the block itself.  `latest_block = None;  latest_block = web3.eth.getBlock('latest')`

Reply:
I don't know how to reassign the variable that the inner function accesses in the outer scope, without using a mutable object or the `nonlocal` keyword, which is unavailable in Python 2.

From the the commit:

> Using a dict for closure around the cache variable, but once we move
to python 3, it can change to storing the block itself, and the inner
function can use the `nonlocal` keyword.

Reply:
Makes complete sense.